staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,1053 @@
+
+	// Create the swap payment timeout timer. If it runs out we cancel the
+	// invoice, but keep monitoring the htlc confirmation.
+	// If the invoice was canceled, e.g. before a restart, we don't need to
+	// set a new deadline.
+	invoice, err := f.cfg.LndClient.LookupInvoice(ctx, f.loopIn.SwapHash)
+	if err != nil {
+		err = fmt.Errorf("unable to look up invoice by swap hash: %w",
+			err)
+
+		return f.HandleError(err)
+	}
+
+	var serverPaymentDeadline time.Timer
+	if invoice.State != invoices.ContractCanceled {
+		// If the invoice is still live we set the timeout to the
+		// remaining payment time. If too much time has elapsed, e.g.
+		// after a restart, we set the timeout to 0 to cancel the
+		// invoice and unlock the deposits immediately.
+		remainingTimeSeconds := f.loopIn.RemainingPaymentTimeSeconds()
+		var timeout time.Duration
+		if remainingTimeSeconds > 0 {
+			timeout = time.Duration(remainingTimeSeconds) *
+				time.Second
+		} else {
+			// If the invoice isn't cancelled yet and the payment
+			// timeout elapsed, we set the timeout to 0 to cancel
+			// the invoice and unlock the deposits immediately.
+			timeout = time.Duration(0) * time.Second
+		}
+		serverPaymentDeadline = *time.NewTimer(timeout)
+	} else {
+		// If the invoice was canceled previously we end our
+		// subscription to invoice updates.
+		cancelInvoiceSubscription()
+	}
+
+	cancelInvoice := func() {
+		f.Errorf("timeout waiting for invoice to be " +
+			"paid, canceling invoice")
+
+		// Cancel the lndclient invoice subscription.
+		cancelInvoiceSubscription()
+
+		err = f.cfg.InvoicesClient.CancelInvoice(ctx, f.loopIn.SwapHash)
+		if err != nil {
+			f.Warnf("unable to cancel invoice "+
+				"for swap hash: %v", err)
+		}
+	}
+
+	for {
+		select {
+		case <-htlcConfChan:
+			f.Infof("htlc tx confirmed")
+
+			htlcConfirmed = true
+
+		case err = <-htlcErrConfChan:
+			f.Errorf("htlc tx conf chan error: %v", err)
+
+		case <-reorgChan:
+			// A reorg happened. We invalidate a previous htlc
+			// confirmation and re-register for the next
+			// confirmation.
+			htlcConfirmed = false
+
+			htlcConfChan, htlcErrConfChan, err = registerHtlcConf()
+			if err != nil {
+				f.Errorf("unable to monitor htlc tx "+
+					"confirmation: %v", err)
+			}
+
+		case <-serverPaymentDeadline.C:
+			// If the server didn't pay the invoice on time, we
+			// cancel the invoice and keep monitoring the htlc tx
+			// confirmation. We also need to unlock the deposits to
+			// re-enable them for loop-ins and withdrawals.
+			cancelInvoice()
+
+			event := f.UnlockDepositsAction(ctx, nil)
+			if event != fsm.OnError {
+				f.Errorf("unable to unlock deposits after " +
+					"payment deadline")
+			}
+
+		case currentHeight := <-blockChan:
+			// If the htlc is confirmed but blockChan fires before
+			// htlcConfChan, we wrongfully reset the deposits in the
+			// code below. Hence, we delay the timeout check here by
+			// one block to ensure that htlcConfChan fires first.
```

Comment:
Thanks for poking here, this comment is outdated. This is the updated comment:
```
// If the htlc is confirmed but blockChan fires before
// htlcConfChan, we would wrongfully assume that the
// htlc tx was not confirmed which would lead to
// returning OnSwapTimedOut in the code below. This in
// turn would prevent us from sweeping the htlc timeout
// path back to us.
// Hence, we delay the timeout check here by one block
// to ensure that htlcConfChan fires first.
```

By the time the htlc is timed out the deposits have already been reset by the payment timeout path.

Do you think this needs more clarification? Thanks again for digging deep.